Model/CheckoutSessionManagement.php

```diff
@@ -720,83 +721,99 @@ private function cancelOrder($order, $quote)
+                    . ' Error: ' . $e->getMessage();
+                $this->logger->error($logError);
                 return [
                     'success' => false,
-                    'message' => $this->getCanceledMessage($amazonSession),
+                    'message' => $this->getTranslationString('Unable to complete Amazon Pay checkout'),
                 ];
             }
-
-            if ($amazonSession['productType'] == 'PayOnly') {
-                if (empty($quote->getCustomerEmail())) {
-                    $quote->setCustomerEmail($amazonSession['buyer']['email']);
-                }
+            if($result['success']) {
+                $orderId = $result['orderId'];
             }
-
-            // get payment to load it in the session, so that a salesrule that relies on payment method conditions
-            // can work as expected
-            $payment = $quote->getPayment();
-
-            // Some checkout flows (especially 3rd party) could get to this point without setting payment method
-            if (empty($payment->getMethod())) {
-                $payment->setMethod(Config::CODE);
+        } else {
+            try {
+                $transaction = $this->getTransaction($amazonSessionId);
+                // If no transaction for amazonSessionId, the order still needs placed (APB)
+                if ($transaction) {
+                    $orderId = $this->magentoCheckoutSession->getLastOrderId();
+                } else {
+                    $result = $this->placeOrder($amazonSessionId);
+                    if ($result['success']) {
+                        $orderId = $result['orderId'];
+                    }
+                }
+            } catch (\Exception $e) {
+                $logError = 'Unable to complete Amazon Pay checkout.'
+                    . ' amazonSessionId: ' . $amazonSessionId . $e->getMessage();
+                $this->logger->error($logError);
+                return [
+                    'success' => false,
+                    'message' => $this->getTranslationString('Unable to complete Amazon Pay checkout'),
+                ];
             }
+        }
 
-            // set amazon session id on payment object to be used in authorize
-            $payment->setAdditionalInformation('amazon_session_id', $amazonSessionId);
-
-            // collect quote totals before placing order (needed for 2.3.0 and lower)
-            // https://github.com/amzn/amazon-payments-magento-2-plugin/issues/992
-            $quote->collectTotals();
-
-            $orderId = $this->cartManagement->placeOrder($quote->getId());
+        if(!$orderId) {
+            $this->logger->error("Unable to complete Amazon Pay checkout. Order Id not found.");
+            return [
+                'success' => false,
+                'message' => $this->getTranslationString('Unable to complete Amazon Pay checkout'),
+            ];
+        }
+        if (empty($amazonSessionId)) {
```

Comment:
I don't think this scenario is possible, as it gets checked as the first if of this method.